### DISCUSSION TITLE
More consistent definition of functions / methods + some suggestions

### DISCUSSION THREAD

**Discussion by mossaiby** [carbon-language/carbon-lang#2026] [Language design]
Hi all,

I have been thinking for a while on ideas similar to what @chandlerc talked about. Even I was thinking on creating a simple language using LLVM for my own needs (number crunching), without all the complexities of C++, but hey, we all know how long it takes...

After watching the intro video, I had some strange feelings about the syntax, and I think many others have that too. I know that you are trying to simplifying things (especially parsing), but I think it has gone a bit too far and became a little too weird for a successor language. Here is my views, with the hope that they help:

1. I think there is some inconsistency in defining the argument and return types for a function / method; the return type uses `->` while every other place is `:` is used (see suggestions below). Good to see that it is like C++ in some sense, but it doesn't match the other parts well.

2. While I completely understand some decisions taken for simplicity of parsing, I think we could do things so that C++ developers feel better; for example, instead of using `fn` keyword, we could elaborate something like C++ way of defining lambdas for functions:
```c++
[] Print(i: i32): void {...}
class test {
...
  [self: Self] Summarize(i: i32): void {...}
...
}
```
For methods / generics, you can still put deduced parameters there. This way you do not need to mimic Rust, and still I think parsing is easy enough.

If this is accepted, I believe you can even revert to C / C++ way of defining types of parameters, return type and variable declaration, without the vexing parse problem:
```c++
[] Print(i32 type, i32 num) -> void {...}
class test {
...
  [Self self] Summarize(i32 i, i32 num) -> void {...}
...
}
i32 test = 5;
[] (i32) -> i32 func; // define a function-type variable; add const / mutable as needed
```

3. The next thing I want to complain about is `var` and `let` keywords. These remind me of JavaScript, except that they have completely different meanings here and there we have `const` as well. A suggestion might be using one of them (`var`? at least it is C#-like!) and use the `const` (for non-mutable) keyword whenever needed. It is much more readable, I think. Even you can use `mutable` keyword for the purpose.

4. The syntax for generics / concepts is too weird... consider using something other than `!` for it. C++ uses `!` for 'not' operator, which seems counterintuitive here.

Thanks!

**Comment by OlaFosheimGrostad** [comment]
Hi, you might want to chime in on issues that cover some of these. 

When we are talking usability then I weight rank legibility/readability/scanning over consistency, within reason.  So for function signatures I would want this priority:
1. What does it do?
2. What does it create?
3. What constraints are relevant for its usage?

So priority 1 means the name should be to the left. Priority 2 means that the return type should be given weight, which can be done with a unique marker such as `->`. Priority 3 means that it should be possible to push deduction parameters down to the next line (issue #1738).

Note that the `->` marker is consistent with C++ `auto f() -> return_type` notation and formal notations.

`name:type` is consistent with formal notation and many languages, it is also better than `type name` as the latter does not line up names for long type specs. In general, what you want to see first as a developer should be pushed all the way to the left (or to the top), so that means the name ought to be to the left. The `:` makes it more visually distinct what is a type and what is a name binding. This is also the order of a function signature, the name is to the left, and the "anonymous" typing part is following to the right.

Carbon has chosen a conservative syntax approach with introducer-keywords in the tradition of ML (or Algol). I personally don't think this redundancy is necessary now that people use IDEs that highlight syntax mistakes. The argument is that it makes tooling easier, but then again, maybe it would be better to provide ready-made parser-libs?

Regarding your point 3: I saw somewhere that @chandlerc is open for doing user testing with other syntax alternatives. If the compiler AST is clean, then it should quite easy to play with multiple syntaxes during initial testing (as long as macro-like rewrites are avoided). Syntax is easy to change, semantics are not so easy to change, so it is more important that the focus is on the latter at this point, IMO.

Regarding your point 4: I have a preference for "and", "or", "not" (C++ also has these) as I think they are less noisy than sigils, so I think Carbon made a good move there.

The biggest syntax issue for me personally is that Carbon has chosen to depart from the all-lowercase naming scheme for language constructs and core standard library constructs  (array etc), this leads to a more noisy screen-display and is clashing with various existing C++ code base style practices.

**Reply by mossaiby** [reply]
Surely I agree with many of your points, but maybe our priorities differ. For example, 

> `name:type` is consistent with formal notation and many languages, it is also better than `type name` as the latter does not line up names for long type specs.

Everybody loves clean and beautiful code. But for a _successor_ language, I think we must prioritize keeping harmless parts of the syntax intact. IMO Carbon, at first look, doesn't seem to be a C++ replacement in this sense. I should confess that the first time, I left watching @chandlerc's intro video, because the syntax was so weird for me. If we can use C++ keywords for something similar in Carbon, we should do that too keep them as similar as possible.

Also, to me, deduction parameters are like template parameters; in C++ we have them before the function / class itself, and usually put them on their own line. This is also possible here. Anyway, I still think that the syntax below has the most of what we want as developers:

```c++
[] Print(i32 type, i32 num) -> void {...}
class test {
...
  [Self self] Summarize(i32 i, i32 num) -> void {...}
...
}
i32 test = 5;
[] (i32) -> i32 func; // define a function-type variable; add const / mutable as needed
```

It also solves the unused function / method parameter I saw on some other discussions.

**Reply by OlaFosheimGrostad** [reply]
FWIW, I think the «successor language» term is a rhetorical phrase that creates a different set of expectations
 than what seems to be the goals for the project. Carbon seems to aim for 99.9% C++ interop and enable a modest level of transpiling/conversion (clearly not full support as Carbon might end up with single inheritance and a different type of error propagation).  It does not makes sense to call Carbon a «successor language» IMHO, just like it doesn't make sense to call languages with good C-interop and C-like capabilities «successor languages» for C.

**Reply by KateGregory** [reply]
I think we've been really clear on what we mean by successor language. The shared ecosystem is key, which means interop is vital. Adopting Carbon doesn't mean leaving behind third party libraries you've come to rely on or your own old code. That's it, that's all. It doesn't mean the syntax is the same or you don't have to learn a new language - you will have to learn a new language. Parts of it will be familiar, and often when there's no real difference among options we choose the one that is more natural to C++ developers.

**Reply by tkoeppe** [reply]
> I think we've been really clear on what we mean by successor language. The shared ecosystem is key, which means interop is vital. Adopting Carbon doesn't mean leaving behind third party libraries you've come to rely on or your own old code. That's it, that's all. It doesn't mean the syntax is the same or you don't have to learn a new language - you will have to learn a new language. Parts of it will be familiar, and often when there's no real difference among options we choose the one that is more natural to C++ developers.

Thanks for this clarification! I've been meaning to post something similar, as there seems to be a common (mis?)understanding of what we mean by "successor language", so maybe here is a good place. As far as I understand it, the term has nothing to do with whether Carbon, or anything, will succeed C++, nor whether any other language could also succeed C++. Rather, it is purely a way to label the design approach so as to distinguish the approach from an _alternative_ that Chandler called "supersetting": If language B supersets language A, then code for A remains valid code for B, and the expected adoption route is to replace the toolchain for A with the toolchain for B and thus enable the new features of B. By contrast, a successor language S for A is a completely different language that requires its own toolchain, and adoption consists of building S code with the new S toolchain, but having that toolchain emit artefacts that integrate seamlessly with the existing A toolchain.

In that light, any expectation that "being a successor language" means that Carbon should retain as much syntax of C++ as possible is opposite to the actual design approach: Retaining syntax is what you would do in a supersetting language, but _not_ in a successor language. A successor language is free to be entirely different, as long as it can be _integrated into the existing ecosystem_.

**Reply by OlaFosheimGrostad** [reply]
> as there seems to be a common (mis?)understanding of what we mean by "successor language"

If the messaging doesn't work, change the message (the term "succession" has very strong connotations, so trying to redefine it is an uphill battle).

**Reply by mossaiby** [reply]
Thanks @KateGregory and @tkoeppe for clarification.

>  Parts of it will be familiar, and often when there's no real difference among options we choose the one that is more natural to C++ developers.

I strongly agree with this point; if we can use _familiar_ things from C++, why we should use _unfamiliar_ ones? So if I can suggest a new syntax which does NOT necessarily compile on C++, but provides the characteristics foreseen for Carbon (ease of parsing, etc.) *AND* looks more familiar to C++ developers, the leads will agree with that, right? I really want to do that, as I see merits in Carbon. Can you please guide me what does not fulfill your expectations in my last post? Maybe we can fix that.

**Reply by YagaoDirac** [reply]
> FWIW, I think the «successor language» term is a rhetorical phrase that creates a different set of expectations than what seems to be the goals for the project. Carbon seems to aim for 99.9% C++ interop and enable a modest level of transpiling/conversion (clearly not full support as Carbon might end up with single inheritance and a different type of error propagation). It does not makes sense to call Carbon a «successor language» IMHO, just like it doesn't make sense to call languages with good C-interop and C-like capabilities «successor languages» for C.

Maybe a better name should be c+++. No need to have a new name imo.

**Reply by OlaFosheimGrostad** [reply]
> Maybe a better name should be c+++. No need to have a new name imo.

I think "Carbon" is ok, but it is difficult to search for. A more unique name could be useful.

( I am not sure if Carbon can be classified as being in the C-family at all, so "C+++" might raise the wrong expectations.)

**Reply by jonmeow** [reply]
Hi @YagaoDirac @OlaFosheimGrostad -- while we appreciate the jokes, they're a bit off-topic and are only serving to distract from the conversation. There's a serious discussion going on, so please don't derail it.
